[WIP] add stream-timeout option to `kubectl exec`

juanvallejo opened the pull request:
**Release note**:
```release-note
adds a --stream-timeout option to "kubectl exec" in order to timeout once inside of a remote shell
```

This patch implements a timeout option for remote streams. This allows a
maximum time to be specified for a remote shell to remain open. If a
`--stream-timeout` value `> 0` is provided, the stream executor will not
wait for the `stdout` and `stderr` streams of a remote shell to finish
copying, but will rather close the stream and exit with error `Timeout
exceeded for this operation`.

Although `kubectl exec` already supports the global
`request-timeout` option as well as the local `--timeout` option, the
former is only for making ther inital request with a `restclient`, and
the latter is the amount of time to wait before a pod is retrieved.
Therefore, it was necessary to introduce a third timeout option for
specifically timing out the command once inside of a remote shell.

**Example**
```
$ kubectl exec -- /bin/sh -c "while true; do echo test; sleep 1s; done"
Defaulting container name to mypodcontainer.
Use 'kubectl describe pod/mypod' to see all of the containers in this pod.
test
test
test
...
^C

$ kubectl exec --stream-timeout=5s mypod -- /bin/sh -c "while true; do echo test; sleep 1s; done"
Defaulting container name to mypodcontainer.
Use 'kubectl describe pod/mypod' to see all of the containers in this pod.
test
test
test
test
test
error: Timeout exceeded for this operation.
```

@kubernetes/cli-review @fabianofranz

juanvallejo committed:
add stream-timeout option to `kubectl exec`

This patch implements a timeout option for remote streams. This allows a
maximum time to be specified for a remote shell to remain open. If a
`--stream-timeout` value `> 0` is provided, the stream executor will not
wait for the `stdout` and `stderr` streams of a remote shell to finish
copying, but will rather close the stream and exit with error `Timeout
exceeded for this operation`.

Although `kubectl exec` already supports the global
`request-timeout` option as well as the local `--timeout` option, the
former is only for making ther inital request with a `restclient`, and
the latter is the amount of time to wait before a pod is retrieved.
Therefore, it was necessary to introduce a third timeout option for
specifically timing out the command once inside of a remote shell.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38642)
<!-- Reviewable:end -->

juanvallejo commented on pkg/kubectl/cmd/exec.go:
```diff
@@ -159,6 +167,8 @@ func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []s
 		}
 	}
 
+	p.StreamTimeout = cmdutil.GetFlagDuration(cmd, "stream-timeout")
```
@smarterclayton or @fabianofranz would it make more sense from a user's perspective to just re-use the value from the global `--request-timeout` flag, or is it worth adding a third timeout option?

fabianofranz replied:
The global `--request-timeout` makes sense to me.

juanvallejo commented:
Related bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1314240

smarterclayton commented:
I'm not sure I would call this `stream-timeout` in general. Is this an idle timeout, or a duration timeout?  request-timeout is duration.

juanvallejo commented:
@smarterclayton per this implementation at least, it is a duration timeout. Regardless of user input, it will exit after the specified time

juanvallejo committed:
update opt to only timeout if idle

Instead of closing the remote shell after the amount of time defined in
--stream-timeout, this patch takes into account any bytes being copied
to remoteStdout. If there are any bytes being copied from the remote
shell's standard output, the timeout counter will be reset back to the
original value provided by the user. The shell will only actually
timeout once the timeout counter reaches zero from the original value
specified by the user.

juanvallejo commented:
@smarterclayton @fabianofranz 
The second commit https://github.com/kubernetes/kubernetes/pull/38642/commits/40f6ad44ab52ed6c10d51ce7132d043fbc2aa97e turns this patch into an "idle timeout". This means that instead of the remote shell closing on a user after a specified timeout value `x`, it will only only close on that user after `x` seconds/units of inactivity.

Inactivity is defined as no bytes being copied from the `remoteStdout` reader to `Stdout` writer.
**Example with oc rsh (also affected by this patch)**
```
$ oc rsh --stream-timeout=5s idling-echo-5-972o2
sh-4.2$ while true; do echo test; sleep 1; done
test
test
test
...
```
The above example will keep running indefinitely, as there is going to be output being copied from `remoteStdout` to `Stdout` until the infinite loop is broken.

```
$ oc rsh --stream-timeout=5s idling-echo-5-972o2
sh-4.2$
# 5 seconds later
error: Timeout exceeded for this operation.
```
Since the remote shell was inactive for `5s`, the timeout actually occurs.

```
$ oc rsh --stream-timeout=5s idling-echo-5-972o2
sh-4.2$
# 4 seconds later
sh-4.2$ ls
# timeout counter reset back to 5s
```
Because there was data to be copied (`ls`) a second before the specified timeout, there will have to be inactivity for another `5s` before the timeout actually occurs.

smarterclayton commented:
Why would I want a duration timeout on exec?

ncdc commented:
We already support idle timeouts for streaming operations (exec, attach, port forwarding), but the value is only configurable by the cluster administrator. If we decide we want the user to have some amount of control, I would prefer that we pass the user-supplied value through to the kubelet and then use min(user supplied value, admin supplied value). I don't think we need to add any code to implement idle timeout client-side.

juanvallejo commented:
@smarterclayton 
> Why would I want a duration timeout on exec?

I convert it to an idle timeout here: https://github.com/kubernetes/kubernetes/pull/38642/commits/40f6ad44ab52ed6c10d51ce7132d043fbc2aa97e (https://github.com/kubernetes/kubernetes/pull/38642#issuecomment-266865954)

@ncdc 
> We already support idle timeouts for streaming operations (exec, attach, port forwarding), but the value is only configurable by the cluster administrator. If we decide we want the user to have some amount of control, I would prefer that we pass the user-supplied value through to the kubelet and then use min(user supplied value, admin supplied value). I don't think we need to add any code to implement idle timeout client-side.

Sounds good, I could turn this PR into something that updates this. Would the value that is configurable by the cluster admin be handled in `pkg/kubelet/server/remotecommand`?

ncdc commented:
The cluster admin configuration option is called StreamingConnectionIdleTimeout

juanvallejo commented:
@ncdc 
> The cluster admin configuration option is called StreamingConnectionIdleTimeout

Thanks, I'll put something together

juanvallejo commented:
@ncdc 
>  If we decide we want the user to have some amount of control, I would prefer that we pass the user-supplied value through to the kubelet and then use min(user supplied value, admin supplied value)

Based on the [original bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=1314240) for the issue addressed by this PR, the user requested a *longer* idle timeout, which means that this will still not be solved if we always choose the min value between `user value` and `cluster config value`. An option could be to still go ahead and implement this, but always have the user's value override the cluster config value. wdyt?

cc @fabianofranz

ncdc commented:
I believe the report in the bz says that the rsh session died after 1 to 2 minutes. The default StreamingConnectionIdleTimeout value is 4 hours. It sounds like something else is going on with that user's environment.

I do *NOT* want to reimplement idle timeouts. I would be open to a discussion about whether or not the user should be allowed to override the cluster admin's setting, although my vote is "no".

juanvallejo commented:
@ncdc @smarterclayton 
closing this pr as the feature request in the bugzilla already exists in the cluster config as `StreamingConnectionIdleTimeout` and the root of the bugzilla issue is not directly caused by this option not existing in the client side.
